### PR TITLE
Feature/add lisk

### PR DESCRIPTION
**Pull request by gost3228**
Add LISK coin to exchanger

### CODE REVIEW

**Comment by adamant-al** [comment]
Hi,
- base_lsk.js better to name lskBaseCoin.js
- LskBaseCoin should extend baseCoin, not btcBaseCoin
- lisk_utils.js better to name lsk_utils.js
- when user have old account, with previous Lisk core address, the bot say like "Exchange Bot [TEST] thinks transaction of 1 LSK is wrong. Sender expected: 1543470903438799018L, but real sender is lskq524hdpsu4qy5gud74yrtrtvbu8ehkmrja6ptr.". It is because both these addresses saved in ADAMANT's key-value storage, and function getKvsCryptoAddress() took first one. Fix this function with adding "orderBy=timestamp:desc" to a request. See: https://endless.adamant.im/api/states/get?key=lsk:address&senderId=U11452494217613574670&orderBy=timestamp:desc
- Wrong confirmations counting. First time tx seen: "Updated Tx 8bcfc29f2c6ed5… confirmations: -1"
- Fix logging "Successfully built Tx cdb0b9de9… to send 0.66822015 LSK to lskq524… with 0.0016 LSK fee: [object Object]".
- Calculate Lisk tx fee. See src/lib/lisk/lsk-base-api.js in adamant-im.

File: helpers/cryptos/lisk_utils.js [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+  /**
+   * Returns Tx status and details from the blockchain
+   * @override
+   * @param {String} txid Tx ID to fetch
+   * @param {Boolean}disableLogging
+   * @return {Object}
+   * Used for income Tx security validation (deepExchangeValidator): senderId, recipientId, amount, timestamp
+   * Used for checking income Tx status (confirmationsCounter), exchange and send-back Tx status (sentTxChecker):
+   * status, confirmations || height
+   * Not used, additional info: hash (already known), blockId, fee, recipients, senders
```
Comment (adamant-al) [inline helpers/cryptos/lisk_utils.js:277]: Are you sure?

File: package.json [hunk 2]
```diff
@@ -37,10 +37,11 @@
   "author": "Aleksei Lebedev, RomanS <devs@adamant.im> (https://adamant.im)",
   "license": "GPL-3.0",
   "dependencies": {
-    "adamant-api": "1.3.0",
+    "adamant-api": "github:Adamant-im/adamant-api-jsclient#feature/add-lisk",
```
Comment (adamant-al) [inline package.json:40]: Use adamant-api 1.4.0

Comment (adamant-al) [reply package.json:40]: Not yet.

**Comment by adamant-al** [comment]
@gost3228 Try to change outdated `"lisk-sdk": "^5.2.1"` to newer one. May be even separate modules like `lisk-transactions` and `lisk-cryptography`.

File: package.json [hunk 3]
```diff
@@ -25,36 +25,38 @@
...
+    "binance coin",
+    "lsk",
+    "lisk"
   ],
   "author": "Aleksei Lebedev, RomanS <devs@adamant.im> (https://adamant.im)",
   "license": "GPL-3.0",
   "dependencies": {
-    "adamant-api": "1.3.0",
-    "axios": "0.23.0",
+    "adamant-api": "github:Adamant-im/adamant-api-jsclient#feature/add-lisk",
```
Comment (adamant-al) [inline package.json:41]: To update before merge

File: .editorconfig [hunk 4]
```diff
@@ -8,8 +8,8 @@ charset = utf-8
 end_of_line = lf
 insert_final_newline = true
 trim_trailing_whitespace = true
-indent_style = tab
-tab_width = 4
+indent_style = space
+tab_width = 2
```
Comment (MerdyNumber1) [inline .editorconfig:12]: indent_size is preferable

File: helpers/cryptos/lskBaseCoin.js [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+module.exports = class LskBaseCoin extends baseCoin {
+  constructor(token) {
+    super();
+    this.token = token;
+    this.account.keys = api[token.toLowerCase()].keys(config.passPhrase);
+    this.account.network = this.account.keys.network;
+    this.account.keyPair = this.account.keys.keyPair;
+    this.account.address = this.account.keys.address;
+    this.account.privateKey = this.account.keys.privateKey;
+    this.account.addressHex = this.account.keys.addressHex;
```
Comment (MerdyNumber1) [inline helpers/cryptos/lskBaseCoin.js:22]: `this.account = Object.assign(this.account, this.account.keys)`?
Or object `this.account.keys` contains additional keys?

File: helpers/cryptos/lskBaseCoin.js [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+ * @return {Object}
+ */
+function createClient(url) {
+  const client = axios.create({ baseURL: url });
+  client.interceptors.response.use(null, (error) => {
+    if (error.response && Number(error.response.status) >= 500) {
+      console.error(`Request to ${url} failed.`, error);
+    }
+    if (error.response && Number(error.response.status) === 404) {
+      if (error.response.data && error.response.data.errors && error.response.data.errors[0] && error.response.data.errors[0].message && error.response.data.errors[0].message.includes('was not found')) {
```
Comment (MerdyNumber1) [inline helpers/cryptos/lskBaseCoin.js:222]: You can simplify
`if (error.response?.data?.errors[0]?.message && error.response.data.errors[0].message.includes('was not found'))`

File: helpers/cryptos/lskBaseCoin.js [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+    this.account.address = this.account.keys.address;
+    this.account.privateKey = this.account.keys.privateKey;
+    this.account.addressHex = this.account.keys.addressHex;
+    this.clients = {};
+
+    this.cache.balance = { lifetime: 30000 };
+    this.cache.lastBlock = { lifetime: 60000 };
+
+    setTimeout(() => this.getBalance().then((balance) => log.log(`Initial ${this.token} balance: ${utils.isPositiveOrZeroNumber(balance) ? balance.toFixed(constants.PRINT_DECIMALS) : 'unable to receive'}`)), 1000);
+    setTimeout(() => this.getLastBlockHeight().then((lastBlockHeight) => log.log(`Last ${this.token} block height: ${utils.isPositiveOrZeroNumber(lastBlockHeight) ? lastBlockHeight : 'unable to receive'}`)), 1000);
```
Comment (martiliones) [inline helpers/cryptos/lskBaseCoin.js:29]: Too long line of code. Split into multiple ones

File: helpers/cryptos/lsk_utils.js [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+  get networkIdentifier() {
+    // Testnet: '15f0dacc1060e91818224a94286b13aa04279c640bd5d6f193182031d133df7c'
+    // Mainnet: '4c09e6a781fc4c7bdb936ee815de8f94190f8a7519becd9de2081832be309a99'
+    const networkIdentifier = '4c09e6a781fc4c7bdb936ee815de8f94190f8a7519becd9de2081832be309a99';
+    return Buffer.from(networkIdentifier, 'hex');
+  }
+  getHeight() {
+    return this._get(`${lskNode}/api/node/info`).then(
+        (data) => {
+          return Number(data.data.height) || 0;
```
Comment (martiliones) [inline helpers/cryptos/lsk_utils.js:59]: `(data) => Number(data.data.height) || 0`

File: helpers/cryptos/lsk_utils.js [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+  }
+
+  /**
+   * Returns last block height of LSK blockchain
+   * @override
+   * @return {Number} or undefined, if unable to get block info
+   */
+  async getLastBlockHeight() {
+    const block = await this.getLastBlock();
+    return block ? block : undefined;
```
Comment (martiliones) [inline helpers/cryptos/lsk_utils.js:90]: `return block || undefined`

File: helpers/cryptos/lsk_utils.js [hunk 10]
```diff
@@ -0,0 +1,291 @@
...
+   */
+  async getBalance() {
+    try {
+      const cached = this.cache.getData('balance', true);
+      if (cached) { // balance is a beddows string or number
+        return this.fromBeddows(cached);
+      }
+      const result = await this._get(`${lskNode}/api/accounts/${this.account.addressHex}`, {});
+      if (result && result.data && (result.data.token.balance !== undefined)) {
+        const balance = result.data.token.balance;
```
Comment (martiliones) [inline helpers/cryptos/lsk_utils.js:106]: `const {balance} = result.data.token`

File: helpers/cryptos/lsk_utils.js [hunk 11]
```diff
@@ -0,0 +1,291 @@
...
+    try {
+      const account = await this._get(`${lskNode}/api/accounts/${this.account.addressHex}`, {});
+      const nonce = Number(account.data.sequence.nonce);
+      fee = this._buildTransaction(params.address, params.value, fee, nonce).minFee;
+      const result = await this.createTransaction(params.address, params.value, fee, nonce);
+      try {
+        if (result) {
+          log.log(`Successfully built Tx ${result.txId} to send ${params.value} ${this.token} to ${params.address} with ${fee} ${this.token} fee.`);
+          const hash = await this.sendTransaction(result.tx);
+          try {
```
Comment (martiliones) [inline helpers/cryptos/lsk_utils.js:205]: what is the 2nd try catch block inside for?

File: helpers/cryptos/lsk_utils.js [hunk 12]
```diff
@@ -0,0 +1,291 @@
...
+   * Returns Tx status and details from the blockchain
+   * @override
+   * @param {String} txid Tx ID to fetch
+   * @param {Boolean} disableLogging Disable logging flag
+   * @return {Object} Formed tx
+   */
+  async getTransaction(txid, disableLogging = false) {
+    return this._getService(`${lskService}/api/v2/transactions/`, { transactionId: txid }).then((result) => {
+      if (typeof result !== 'object') return undefined;
+      if (result && result.data[0]) {
```
Comment (martiliones) [inline helpers/cryptos/lsk_utils.js:284]: remove the `if` statement above and just use `typeof result === 'object'` && result.data[0]`

File: helpers/utils.js [hunk 13]
```diff
@@ -328,4 +328,20 @@ module.exports = {
...
+   * Converts a bytes array to the respective string representation
+   * @param {Array<number>|Uint8Array} bytes bytes array
+   * @return {string}
+   */
+  bytesToHex(bytes = []) {
+    const hex = [];
+
+    bytes.forEach((b) => {
+      hex.push((b >>> 4).toString(16));
+      hex.push((b & 0xF).toString(16));
```
Comment (martiliones) [inline helpers/utils.js:341]: can be used one `push` instead. 
```
hex.push(
  (b >>> 4).toString(16),
  (b & 0xF).toString(16)
)
```

Comment (martiliones) [reply helpers/utils.js:341]: and so the arrow function brackets can be removed

**Review by martiliones — CHANGES_REQUESTED** [review]

**Review by martiliones — APPROVED** [review]

**Review by adamant-al — APPROVED** [review]

**Comment by adamant-al** [comment]
@MerdyNumber1 It's your turn.

**Comment by MerdyNumber1** [comment]
Excellent

**Review by MerdyNumber1 — APPROVED** [review]

**Pull request merged**